test/CrateIntegrationTest/PDO/PDOStatementTest.php

```diff
@@ -348,6 +348,27 @@ public function testInsertNull()
 
     }
 
+    /**
+     * Verify support for CrateDB bulk-operations endpoint.
+     * https://crate.io/docs/crate/reference/en/5.2/interfaces/http.html#bulk-operations
+     */
+    public function testInsertBulk()
+    {
+        // Insert records in bulk mode.
+        $this->pdo->setAttribute(PDO::CRATE_ATTR_BULK_MODE, true);
+        $parameters = [[5, 'foo', 1], [6, 'bar', 2], [7, 'foo', 3], [8, 'bar', 4]];
+        $statement = $this->pdo->prepare('INSERT INTO test_table (id, name, int_type) VALUES (?, ?, ?)');
+        $this->assertTrue($statement->execute($parameters));
+
+        // Verify records have been inserted correctly.
+        $this->pdo->setAttribute(PDO::CRATE_ATTR_BULK_MODE, false);
+        $this->pdo->exec("REFRESH TABLE test_table");
+        $statement = $this->pdo->prepare("SELECT id, name, int_type FROM test_table");
+        $results = $statement->fetchAll(PDO::FETCH_NUM);
+        $this->assertEquals([5, 'foo', 1], $results[0]);
+        $this->assertEquals([8, 'bar', 4], $results[3]);
+    }
+
```

Comment:
.... however, it turned out that it has a usability glitch in its current form. Because bulk operations are only applicable to DML operations (`INSERT`, `UPDATE`, `DELETE`), and deliver a response of a different shape, they are not suitable for `SELECT`-type statements.

Because the `CRATE_ATTR_BULK_MODE` setting is currently handled on the PDO instance level, it currently need to be switched back to non-bulk mode, in order to conduct `SELECT` statements properly.

That's obviously not a good way to work with that mode, so I will try to push back this feature to be enabled on the `PDOStatement`-level only, by using its `$options` dictionary and the corresponding `bulkMode` slot, probably removing `CRATE_ATTR_BULK_MODE` completely again.

Reply:
I've dissolved `PDO::CRATE_ATTR_BULK_MODE` again with c7d4b3bbc7. CrateDB bulk operations will exclusively be available by creating a dedicated `PDOStatement` instance, with `->prepare($sql, array("bulkMode" => true))`, as outlined in the documentation example. This interface is [modeless](https://en.wiktionary.org/wiki/modeless), and as such simpler and better.

-- https://crate-pdo--143.org.readthedocs.build/en/143/connect.html#bulk-operations